### ISSUE TITLE
Server-side async test failures not showing in reporter

### ISSUE THREAD

**Issue by loirit** [practicalmeteor/meteor-mocha#11]
The following line is being printed to the console:

``
(STDERR) MochaRunner.Meteor.publish.mochaServerRunEvents: failures: 1
``

But nothing is showing up in the HTML reporter.

**Comment by jsep** [comment]
@loirit thanks for reporting the issue we will take a look

**Comment by loirit** [comment]
For what it's worth, I was able to work around this by wrapping the assertions in a try-catch block and passing any errors to `done`.

**Comment by tmeasday** [comment]
I think the issue happens when you run fiber-y meteor code. See for instance: https://github.com/tmeasday/meteor-mocha-fibers-problem.git

**Comment by tmeasday** [comment]
(Note that to make it happen you need to force a hot code reload for some reason)

**Comment by rbabayoff** [comment]
@tmeasday thanks for the feedback. Your workaround is actually a standard practice that should be used with all testing frameworks in async tests, most of them will otherwise not know if an exception was raised in an async callback of a test. Some testing frameworks install an uncaught exception handler to deal with uncaught errors in async callbacks. Mocha is one of them, from what I know. But I think we should replace it with a fibers aware one. I'll look into it.

**Comment by tmeasday** [comment]
Hi. I'm not sure that make sense though--There's no actual "async" involved in [my example](https://github.com/tmeasday/meteor-mocha-fibers-problem/blob/master/packages/tester/tester.js) (modulo fibers).

It should appear (from the perspective of mocha anyway) exactly the same as if the sleep wasn't there (in which case it works fine). The exception should bubble up the stack to mocha somewhere -- and it does appear to work the first time you run it, which shows that it does, and indicates the issue is a subtle one, and not irresolvable.

I actually suspect the issue is in the reporter publishing the error across, rather than in mocha core, but it's kind of hard to understand how the code moves through the package.

**Comment by tmeasday** [comment]
I mean theoretically, if all you need to do is catch errors and hand them to the done call back, could the `it()` helper just be reimplemented as:

```js
const oldIt = it;
it = function(fn) {
  oldIt((done) => {
    try {
      fn(done);
      if (fn.length === 0) {
        done();
      }
    } catch(e) {
      done(e);
    }
   });
}
```

But I don't think that's really the issue..?

**Comment by tmeasday** [comment]
Hey @rbabayoff - how do I run the package test suite (the tests in `test-package/mocha-tests.coffee`?) I'd like to add a test demonstrating the issue (which could help in getting to the bottom of it)..

**Comment by tmeasday** [comment]
Ok, I think I sorted it out: https://github.com/tmeasday/meteor-mocha-core/commit/d1e056b00ee7812aa22dc4d832e8024068ff0767

I want to check with @stubailo that my changes make sense but will be sending PRs upstream soon.

Still don't know the exact explanation of the behaviour I was seeing with reload

**Comment by tmeasday** [comment]
Ok, couldn't let it rest, figured out what the explanation was. It was a combination of things.

1. mocha-core (https://github.com/practicalmeteor/meteor-mocha/issues/11#issuecomment-154902071) wasn't properly catching exceptions thrown after a fiber yield and thus was throwing them on the process. This would ordinarily crash the entire server.

2. However, mocha also has a global uncaught exception handler (https://github.com/mochajs/mocha/blob/master/lib/runner.js#L749) which matches such exceptions to the currently running test.

3. *However*, mocha isn't really designed to run the same test suite more than once (https://github.com/mochajs/mocha/issues/995) and this code path (through the uncaught exception handler) triggered a bug in this behaviour (`test.state` was already set to `failed` so the test never `end`-ed). 

So the simple solution is to properly throw the exception as above. However it's perhaps concerning that re-running the tests on client HCR (or simple browser refresh) could cause other issues of the same nature.

**Comment by rbabayoff** [comment]
@tmeasday , that's amazing. Thanks so much for figuring it out. We'll check it internally tomorrow and merge. I guess you figured out how to run the test-package, but if not (from repo root):

```
cd meteor/test-package
meteor test-packages --driver-package=practicalmeteor:mocha ./ 
```

This package includes all types of tests, passing, failing, skipped, etc.

I guess to test the hot code push scenario, a functional test using selenium would probably be a good idea, I'll see when we can get to it.

**Comment by rbabayoff** [comment]
Sorry, just saw PR is not there yet. Will wait for your PR.

**Comment by tmeasday** [comment]
Hi @rbabayoff I'm going to send a PR through to the upstream package (@rissem's), when I have a chance. I think my latest version of the code is more or less correct, just need to squash it and send the PR

**Comment by tmeasday** [comment]
Ok, it seems like there's a complicated mess of different versions of this `server.js` file so I'm not sure where to send the PR.

I'm beginning to wonder why this is all so complicated anyway. Couldn't we just do this?

```
Fiber(function() {
  Mocha.run();
}).run();
```

Does mocha run stuff in timeouts? Why do the it-blocks themselves need to be wrapped?

**Comment by rbabayoff** [comment]
1. If the it ones are not wrapped, then in async tests meteor complains there are no fibers (if I recall correctly). Also, in mocha you can specify test timeouts, which present another issue, fibers wise. Welcome to the messy world of trying to wrap complex async libraries for meteor :-)

2. We cloned rissem's package, fixed some bugs and added some stuff, such as support for displaying server side test source code, so you should work off and create the PR here:

https://github.com/practicalmeteor/meteor-mocha-core

**Comment by tmeasday** [comment]
Hi @rbabayoff,

Ok, you are right; test timeouts would definitely be a problem.

So, `bindEnvironment` (rather than a simple fiber approach, ala https://github.com/practicalmeteor/meteor-mocha-core/compare/practicalmeteor...tmeasday:practicalmeteor) is only helpful if it is called from inside a method (in which case it makes `Meteor.userId()` and co works).

1. So I don't think it makes sense to use it in an `it()` function (i.e. at definition-time), which wouldn't be called from a method. 

2. I don't think there's any reason to use any kind of fiberizing for a describe block (unless they can be called async too? I think not?). Still if we were going to, the same approach for `it()` would make sense.

3. Given the above, I don't think there's any need for `moddedBindEnvironment` which makes this file much simpler.

When you have a chance, can you let me know your thoughts on the above. I'll send a PR simplifying things if you agree.

PS Also, in the mocha package, I noticed a few points of doing:

```
bindEnvironment(function() {
   // do something
})();
```

For the reason above, I don't think there's any purpose in that, beyond setting up a fiber -- but I think you do it inside a publication, which has a fiber anyway.

**Comment by rissem** [comment]
@tmeasday thanks for looking into this. I agree w/ everything in your comment above. And I apologize for creating a confusing mess here that you had to untangle. I mentioned this on another issue, but we should try to unify the mocha-core packages. I'm happy for the official mocha-core package to be https://github.com/practicalmeteor/meteor-mocha-core/. @rbabayoff thoughts? How much has this package diverged from mine?

**Comment by rbabayoff** [comment]
@rissem I just created a PR to your core, so you can compare the changes https://github.com/rissem/meteor-mocha-core/pull/1

Yes, I also think that merging the 2 would be a great idea. If your OK with it, let's let @tmeasday create his PR into our mocha-core and then both @jsep, our lead developer on this, and you can check it on all the packages both of us have that already use mocha for tests. 

I'll also have @jsep also look into how to solve the test timeout issue (if there still is one) - we're thinking of using only the client side mocha version for both meteor client and server side, so we can hack into it easily. We did the same for chai and sinon, no need to use the npm packages, since the browserified versions include all that is needed, for server side too.

Sounds like a plan?

**Comment by tmeasday** [comment]
Do you guys want to take it to a sensible point and then I'll take a look at it? I think that probably makes the most sense as I'm not completely sure of what all the requirements are.

**Comment by rbabayoff** [comment]
I recommend @jsep integrates your suggestions, and we'll see if nothing broke.

**Comment by apendua** [comment]
@jsep Any progress on this one?

**Comment by rbabayoff** [comment]
@apendua , yes, it should be solved in latest version 2.1.0_7.

Let me know if it works.

**Comment by rbabayoff** [comment]
@tmeasday I think this is ready for prime time, including running mocha tests from the command line with  [spacejam](https://www.npmjs.com/package/spacejam).

We've integrated your suggestions, and also solved the uncaught exceptions stuff, both server and client side, allowing mocha to report them.

Let me know if this works for you.

**Comment by apendua** [comment]
@rbabayoff Cool, thanks! Are you planning to release it anytime soon?

**Comment by rbabayoff** [comment]
@apendua already released, v 2.1.0_7, update your dependency.

**Comment by apendua** [comment]
@rbabayoff Nice, thanks!

**Comment by tmeasday** [comment]
Thanks @rbabayoff, I'll update the dependency in the todos app and try it out.

**Comment by apendua** [comment]
@rbabayoff It looks like updating to `2.1.0_7` has not solved the issue for me.

**Comment by rbabayoff** [comment]
@apendua ok, we'll take another look at it. Can you provide a simple package repo that reproduces it?

**Comment by apendua** [comment]
@rbabayoff Actually, I think I can see where the issue is right away. [Here](https://github.com/practicalmeteor/meteor-mocha-core/blob/practicalmeteor/server.js#L33) you should take into account the fact that the return value may be a promise.

**Comment by apendua** [comment]
@rbabayoff Here's a reference implementation that works properly:

https://github.com/anticoders/gagarin/blob/master/lib/mocha/interface.js#L326-L363

**Comment by 0o-de-lally** [comment]
I'm still getting this error. Meteor methods cannot be tested. It will fail in console, but the reporter will show passing. Unless I'm doing something wrong.

```
describe('my module', function () {

  it('does something that should be tested', function (done) {
    Meteor.call('anyMethod', function(err, res){
      chai.assert.equal( 1 ,  2 )
      done()
    })
  })

})
```

>  Exception in delivering result of invoking 'anyMethod': AssertionError: expected 1 to equal 2

And in browser:  "Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test."

**Comment by rbabayoff** [comment]
@keyscores , @apendua and @loirit can you check if this still happens with latest version? We solved some related issues, and we can't reproduce this one.

**Comment by 0o-de-lally** [comment]
@rbabayoff Yes I'm still getting false positives. 
In console I see:

> (STDERR) MochaRunner.runServerTests: failures: 1

In reporter I see a checkmark, for passing.

**Comment by jsep** [comment]
@keyscores can you provide a repo? That would be awesome! :smile:

**Comment by keyscores** [comment]
Certainly:  https://github.com/keyscores/meteor-mocha-issue-repro

It also includes a bonus issue, where the test runner reports a timeout instead of an assertion failure on callbacks.

**Comment by 0o-de-lally** [comment]
Certainly: https://github.com/keyscores/meteor-mocha-issue-reproduction

I've made several variations on sync/async methods, and using fibers/future for waiting for the result. The issue is clearly with using futures in the mocha scope.

I added a bonus issue, where mocha reports a timeout instead of assertion fail within method callbacks.

Hope it helps.

**Comment by jsep** [comment]
Thanks @keyscores for the repo, the fibers are messing around.

I fails even using `Meteor.setTimeout()` to report an uncaught exception, instead it reports timeout error for example:

```
it("Meteor.setTimeout fails to report error", function(done){
 Meteor.setTimeout(function(){
    throw new Error("I'm failling :D!") 
 }, 1000)
})
```

Using 'normal' `setTimeout` it works as expected:

```
it("setTimeout report error", function(done){

 setTimeout(function(){
    throw new Error("I'm failling :D!") 
 }, 1000)
})
```

**Comment by 0o-de-lally** [comment]
Yeah looks like it's any use of fibers that causes it. Any thoughts on the second issue, which is where assertion failures are marked as timeouts? Does that need another github issue?

**Comment by tmeasday** [comment]
1. The issue here: https://github.com/keyscores/meteor-mocha-issue-reproduction/blob/master/server/main.app-test.js#L25
   
   is that `Meteor.call` wraps exception handling in the callback. Possibly we could add a test hook to livedata that passes such exceptions right through. I don't think this is a problem with this package though, it's a core Meteor issue/feature.
2. The issue here is you call `done()` [here](https://github.com/keyscores/meteor-mocha-issue-reproduction/blob/master/server/main.app-test.js#L42). The test has already passed before the exception is thrown. Remove the `done()` and it works as expected
3. Same [here](https://github.com/keyscores/meteor-mocha-issue-reproduction/blob/master/server/main.app-test.js#L88)
4. This failure is the same a 1. https://github.com/keyscores/meteor-mocha-issue-reproduction/blob/master/server/main.app-test.js#L115

**Comment by tmeasday** [comment]
@jsep the `Meteor.setTimeout` problem is the same -- Meteor doing it's own exception handling, which bypasses Mocha's

**Comment by jakelacey2012** [comment]
Hi @tmeasday thanks for you hard work on this it is appreciated! :)

I have a very simple test here, 

``` javascript
import { Meteor } from 'meteor/meteor';
import { chai } from 'meteor/practicalmeteor:chai';
import '../index';      // this imports the meteor methods.... (let me know if you want to see it)

describe('MethodCall', () => {
  if (Meteor.isServer) return;
  it('Should return else', () => {
    Meteor.call('somthing.else', 'else', (err, res) => {
      chai.assert.equal(res, 'somthingelse', 'should be else');
    });
  });
});
```

I get this false positive...

```
I20160918-17:51:25.671(1)? stdout:
I20160918-17:51:25.714(1)? Exception in delivering result of invoking 'somthing.else': assert@http://localhost:5000/packages/practicalmeteor_chai.js?hash=8fb9a4aa7055fa02f8b35e05fb1f05c8787c8a46:900:67
I20160918-17:51:25.715(1)? equal@http://localhost:5000/packages/practicalmeteor_chai.js?hash=8fb9a4aa7055fa02f8b35e05fb1f05c8787c8a46:2635:16
I20160918-17:51:25.715(1)? http://localhost:5000/app/app.js?hash=a033c2a03ca4d25d14dea0cc4d142a94587b3e17:163:24
I20160918-17:51:25.716(1)? http://localhost:5000/packages/meteor.js?hash=ae8b8affa9680bf9720bd8f7fa112f13a62f71c3:1105:27
I20160918-17:51:25.719(1)? _maybeInvokeCallback@http://localhost:5000/packages/ddp-client.js?hash=27502404fad7fc072e57e8b0b6719f40d92709c7:3557:21
I20160918-17:51:25.719(1)? receiveResult@http://localhost:5000/packages/ddp-client.js?hash=27502404fad7fc072e57e8b0b6719f40d92709c7:3577:30
I20160918-17:51:25.720(1)? _livedata_result@http://localhost:5000/packages/ddp-client.js?hash=27502404fad7fc072e57e8b0b6719f40d92709c7:4742:22
I20160918-17:51:25.720(1)? onMessage@http://localhost:5000/packages/ddp-client.js?hash=27502404fad7fc072e57e8b0b6719f40d92709c7:3385:28
I20160918-17:51:25.721(1)? http://localhost:5000/packages/ddp-client.js?hash=27502404fad7fc072e57e8b0b6719f40d92709c7:2736:19
I20160918-17:51:25.721(1)? forEach@[native code]
I20160918-17:51:25.722(1)? forEach@http://localhost:5000/packages/underscore.js?hash=27b3d669b418de8577518760446467e6ff429b1e:149:18
I20160918-17:51:25.722(1)? onmessage@http://localhost:5000/packages/ddp-client.js?hash=27502404fad7fc072e57e8b0b6719f40d92709c7:2735:15
I20160918-17:51:25.722(1)? dispatchEvent@http://localhost:5000/packages/ddp-client.js?hash=27502404fad7fc072e57e8b0b6719f40d92709c7:175:27
I20160918-17:51:25.723(1)? _dispatchMessage@http://localhost:5000/packages/ddp-client.js?hash=27502404fad7fc072e57e8b0b6719f40d92709c7:1160:23
I20160918-17:51:25.723(1)? _didMessage@http://localhost:5000/packages/ddp-client.js?hash=27502404fad7fc072e57e8b0b6719f40d92709c7:1218:34
I20160918-17:51:25.724(1)? onmessage@http://localhost:5000/packages/ddp-client.js?hash=27502404fad7fc072e57e8b0b6719f40d92709c7:1365:28
I20160918-17:51:28.471(1)? All client and server tests finished!
```

**Comment by tmeasday** [comment]
@jakelacey2012 I think the thing to do here is to wrap your `Meteor.call` callbacks in an exception handler:

``` js
  it('Should return else', (done) => {
    Meteor.call('somthing.else', 'else', (err, res) => {
      try {
        chai.assert.equal(res, 'somthingelse', 'should be else');
        done();
      } catch (e) { done(e); }
    });
  });
```

In general you have to do this for mocha when you are doing async stuff anyway (unless you can return a promise), I often have something like:

``` js
function wrap(done, cb) {
   return (...args) => {
      try { return cb(...args); } catch (e) { done(e); }
   }
}

it('Should return else', () => {
  Meteor.call('somthing.else', 'else', wrap(done, (err, res) => {
    chai.assert.equal(res, 'somthingelse', 'should be else');
  }));
});
```

**Comment by tmeasday** [comment]
Unless others have a better way of dealing with this?

**Comment by jakelacey2012** [comment]
@tmeasday cheers will give that a go :)

**Comment by jakelacey2012** [comment]
sweet, that worked for me. The only thing I needed to add was calling the done function inside the callback statement

``` javascript
const wrap = (done, cb) => {
  return (...args) => {
      try { return cb(...args); } catch (e) { done(e); }
   }
}

describe('MethodCall', () => {

  if (Meteor.isServer) return;

  it('Should return else', (done) => {

    Meteor.call('somthing.else', 'else', wrap(done, (err, res) => {
      chai.assert.equal(res, 'somthing else else', 'should be else');
      done(); <!---- I did need to add this, but I can see you did that in the other example so I will let you off this time :) 
    }));

  });

});
```

**Comment by macsj200** [comment]
I'm still getting this issue. It seems to work fine on the first run, but after a hot code push reload the server side test results just aren't delivered to the UI. Any ideas?

**Comment by apendua** [comment]
@macsj200 You can try another - yet very similar - test driver:

https://github.com/gagarinjs/meteor-mocha

**Comment by macsj200** [comment]
I've evaluated some other driver packages, and many of them have their own issues. I'd like to stick with the packages suggested in the meteor guide. That being said, I'm willing to consider this as a workaround if all else fails. Do we have any idea what's causing the core issue here? I can help make a minimal repro with some guidance.

Edit: tried the suggested package, worked like a dream. I guess I'll just go with that option for now. Thanks for the help!
